src/config.rs

```diff
@@ -68,6 +73,72 @@ impl Config {
     }
 }
 
+#[derive(Serialize, Deserialize)]
+pub struct SubProcess {
+    pub command: String,
+    pub working_dir: Option<String>,
+    #[serde(default)]
+    pub keep_on_exit: bool,
+    #[serde(default)]
+    pub stdout: bool,
+    #[serde(default)]
+    pub stderr: bool,
+    #[serde(default)]
+    pub foreground: bool,
+}
+
+impl SubProcess {
+    pub fn spawn(&self, tmp_dir: &Path) -> Option<process::Child> {
+        let current_dir = if let Some(working_dir) = &self.working_dir {
+            Path::new(working_dir)
+        } else {
+            tmp_dir
+        };
+
+        let mut process = process::Command::new(run::get_shell());
+
+        process.current_dir(&current_dir);
+
+        #[cfg(unix)]
+        process.arg("-c");
+        #[cfg(windows)]
+        process.arg("/k");
+
+        if !self.stdout {
+            process.stdout(process::Stdio::null());
+        }
+
+        if !self.stderr {
+            process.stderr(process::Stdio::null());
+        }
+
+        process.arg(self.command.clone());
+
+        if self.foreground {}
+
+        if !self.foreground {
+            match process.arg(self.command.clone()).spawn().ok() {
+                Some(child) => Some(child).filter(|_| !self.keep_on_exit && self.foreground),
```

Comment:
There is no point to evaluate something using `self.foreground` at this point because we already ensure that this foreground variable is `false` at the beginning of the block:

```rust
if !self.foreground {
```

```suggestion
        if !self.foreground {
            match process.arg(self.command.clone()).spawn().ok() {
                Some(child) => Some(child).filter(|_| !self.keep_on_exit && false),
```

From there you can refactor the boolean expression:

`anything && false` == `false`

```suggestion
        if !self.foreground {
            match process.arg(self.command.clone()).spawn().ok() {
                Some(child) => Some(child).filter(|_| false),
```

Then finally:

```suggestion
        if !self.foreground {
            match process.arg(self.command.clone()).spawn().ok() {
                Some(_child) => None,
```

Reply:
If i use `Some(child) => None`, i will not kill the subprocess while cleaning up. I need the `!self.keep_on_exit`

Reply:
Oh you can do that!

But if you keep `&& self.foreground` it's like writing `&& false` which cancelled everything you did.

I was merely refactoring your wrong code to show you it's not doing what you think it does.